### ISSUE TITLE
Pass Roslyn diagnostic id as VSCode diagnostic code

### ISSUE THREAD

**Issue by henryju** [OmniSharp/omnisharp-roslyn#1622]
Hi,

A bit of context: SonarSource is providing a C# analyzer based on Roslyn. We already have a VSCode extension (SonarLint) with our own language server, that is providing analysis for other languages.

For C#, we plan to rely on you :) we will ensure our analyzer is used as part of the build (similar to what Roslynator is doing). Still we would like to provide some extra features, like some code actions. The problem we see is that currently, diagnostics created by omnisharp-vscode have no source nor code attribute set. This is a problem for us to differentiate diagnostics that are coming from our analyzer, from other one (compiler, or other third party analyzer).

I noticed that the diagnostic id is in fact appended to the message very early: https://github.com/OmniSharp/omnisharp-roslyn/blob/57e8052bc905a77a74c5f7623f6715df4aac4e47/src/OmniSharp.Roslyn.CSharp/Helpers/DiagnosticExtensions.cs#L25

Is there any reason for that? Would you be open to a pullrequest where the diagnostic id is instead passed through the VSCode diagnostic code?

Regarding the VSCode diagnostic source, it seems Roslyn itself is not exposing the information to know from which analyzer a diagnostic is coming (it would have been great to have the analyzerId). So maybe we could hardcode "csharp", "roslyn", or "omnisharp", just to avoid mixing diagnostics you create with something else. I can also provide a PR if you like.

**Comment by filipw** [comment]
thanks for the issue. I am not sure I understand what you are trying to do.

Diagnostic IDs should be globally unique so you should always be able to understand which diagnostics come from where i.e. there are *IDExxx* ones (which, I guess, are the "roslyn" diagnostics you mentioned), *CSxxx* ones (which, I guess are the "csharp" diagnostics you mentioned).OmniSharp uses no other built-in analyzers. Now, if you have Roslynator installed, that uses *RCSxxx* and so on. 

wouldn't your diagnostics simply have different prefix?

Also, I am not sure I follow how you want to integrate - I moved this to omnisharp-roslyn, since it appears to me you want to integrated the analyzers into the server side, since I guess you'd want to it to invoke your code fixes too, but I don't know if that is the correct assumption.

**Comment by henryju** [comment]
> However, I am not sure I understand what you are trying to do.

We want to provide code actions for our diagnostics, using our existing language server. In the current state, here is an example of what our language server receive as code action request, when triggered on a .cs file having an issue reported by omnisharp-vscode + our analyzer:

```
[Trace - 11:33:21] Sending request 'textDocument/codeAction - (10)'.
Params: {
    "textDocument": {
        "uri": "file:///home/julien/tmp/sample-dotnetcore/MyConsoleApp/Program.cs"
    },
    "range": {
        "start": {
            "line": 11,
            "character": 21
        },
        "end": {
            "line": 11,
            "character": 21
        }
    },
    "context": {
        "diagnostics": [
            {
                "range": {
                    "start": {
                        "line": 11,
                        "character": 19
                    },
                    "end": {
                        "line": 11,
                        "character": 23
                    }
                },
                "message": "Complete the task associated to this 'TODO' comment. (S1135) [MyConsoleApp]",
                "severity": 2
            }
        ]
    }
}
```

As you can see, in this (VSCode) diagnostic, the `code` and `source` attributes are missing, making it hard for our language server to return the appropriate code actions.
I noticed that you are appending the diagnostic id to the message, and I was proposing to change that, in order to have:

```
[Trace - 11:33:21] Sending request 'textDocument/codeAction - (10)'.
Params: {
    "textDocument": {
        "uri": "file:///home/julien/tmp/sample-dotnetcore/MyConsoleApp/Program.cs"
    },
    "range": {
        "start": {
            "line": 11,
            "character": 21
        },
        "end": {
            "line": 11,
            "character": 21
        }
    },
    "context": {
        "diagnostics": [
            {
                "range": {
                    "start": {
                        "line": 11,
                        "character": 19
                    },
                    "end": {
                        "line": 11,
                        "character": 23
                    }
                },
                "message": "Complete the task associated to this 'TODO' comment. [MyConsoleApp]",
                "code": "S1135",
                "severity": 2
            }
        ]
    }
}
```

That would already be a nice improvement for us. Else we would have to parse the diagnostic message, assuming you will never change the format...

Regarding the `source` attribute, I know that Roslyn diagnostic IDs are supposed to be globally unique, but that's not necessarily true for the entire VSCode ecosystem. You could imagine another VSCode extension providing some diagnostics for .cs files (spellchecker, ...), and there is a (quite small I admit) chance of conflict. That's why IMO there is this `source` attribute, used to identify the tool at the origin of the diagnostic. My understanding of the [VSCode API](https://code.visualstudio.com/api/references/vscode-api#Diagnostic) is that it is safe to consider diagnostic code to be unique for a given source. That's why I'm suggesting to simply hardcode it to "omnisharp", or "roslyn", depending on what you consider to be the source of those diagnostics.

In the end, I would like to have my language server receiving:

```
[Trace - 11:33:21] Sending request 'textDocument/codeAction - (10)'.
Params: {
    "textDocument": {
        "uri": "file:///home/julien/tmp/sample-dotnetcore/MyConsoleApp/Program.cs"
    },
    "range": {
        "start": {
            "line": 11,
            "character": 21
        },
        "end": {
            "line": 11,
            "character": 21
        }
    },
    "context": {
        "diagnostics": [
            {
                "range": {
                    "start": {
                        "line": 11,
                        "character": 19
                    },
                    "end": {
                        "line": 11,
                        "character": 23
                    }
                },
                "message": "Complete the task associated to this 'TODO' comment. [MyConsoleApp]",
                "code": "S1135",
                "source": "omnisharp",
                "severity": 2
            }
        ]
    }
}
```

Does it make sense?

> wouldn't your diagnostics simply have different prefix?
yes, Sxxxx, but here the point is to have a reliable way to get the id

> it appears to me you want to integrated the analyzers into the server side
what we want is to run our analyzer as a third party analyzer, like mentioned in [this blog](https://www.strathweb.com/2019/04/roslyn-analyzers-in-code-fixes-in-omnisharp-and-vs-code/), ideally without too much end user manual configuration (but that's a different story, keep it for another topic ;))

**Comment by henryju** [comment]
To make my request (hopefully) clearer, I submitted 2 pull requests with the changes related to the diagnostic code proper support.

**Comment by filipw** [comment]
implemented in #1623
